File: pkg/rtweekend/perlin.go
package rtweekend

import (
	"math"
)

type Perlin struct {
	ranvec              []Vec3
	permX, permY, permZ []int
}

func (*Perlin) POINT_COUNT() int {
	return 256
}

/* Return pointer of Perlin object
 */
func NewPerlin() *Perlin {
	perlin := new(Perlin)

	perlin.ranvec = make([]Vec3, perlin.POINT_COUNT())
	perlin.permX = perlin.perlinGeneratePerm()
	perlin.permY = perlin.perlinGeneratePerm()
	perlin.permZ = perlin.perlinGeneratePerm()

	for i := 0; i < perlin.POINT_COUNT(); i++ {
		perlin.ranvec[i] = *UnitVector(RandomVec3InRange(-1, 1))
	}

	return perlin
}

/* Output random noise
   Param
       p_ *Point3  Coordinate to add noise
*/
func (p *Perlin) Noise(p_ *Point3) float64 {
	u := p_.X() - math.Floor(p_.X())
	v := p_.Y() - math.Floor(p_.Y())
	w := p_.Z() - math.Floor(p_.Z())
	i := int(math.Floor(p_.X()))
	j := int(math.Floor(p_.Y()))
	k := int(math.Floor(p_.Z()))
	var c [2][2][2]Vec3

	for di := 0; di < 2; di++ {
		for dj := 0; dj < 2; dj++ {
			for dk := 0; dk < 2; dk++ {
				c[di][dj][dk] = p.ranvec[p.permX[(i+di)&255]^
					p.permY[(j+dj)&255]^
					p.permZ[(k+dk)&255]]
			}
		}
	}

	return p.trilinearInterp(&c, u, v, w)
}

func (perlin *Perlin) Turb(p Point3, depth ...int) float64 {
	var depth_ int
	accum := 0.0
	weight := 1.0

	if len(depth) == 1 {
		depth_ = depth[0]
	} else {
		depth_ = 7
	}

	for i := 0; i < depth_; i++ {
		accum += weight * perlin.Noise(&p)
		weight *= 0.5
		p.MultToThis(2)
	}

	return math.Abs(accum)
}

// Generate Perlin noise
func (p *Perlin) perlinGeneratePerm() []int {
	retval := make([]int, p.POINT_COUNT())

	for i := 0; i < p.POINT_COUNT(); i++ {
		retval[i] = i
	}

	p.permute(&retval, p.POINT_COUNT())

	return retval
}

/* Permute data in Perlin
   p_  *[]int  Data to permute
   n   int     Size of data
*/
func (p *Perlin) permute(p_ *[]int, n int) {
	for i := n - 1; i > 0; i-- {
		target := RandomIntInRange(0, i)
		tmp := (*p_)[i]
		(*p_)[i] = (*p_)[target]
		(*p_)[target] = tmp
	}
}

/* Trilinear inpolation
   Params
       c *[2][2][2]float64 3D array to calculate interpolation
       u float64           Gain for x axis
       v float64           Gain for y axis
       w float64           Gain for z axis
*/
func (p *Perlin) trilinearInterp(c *[2][2][2]Vec3, u, v, w float64) float64 {
	accum := 0.0
	uu := u * u * (3 - 2*u)
	vv := v * v * (3 - 2*v)
	ww := w * w * (3 - 2*w)

	for i := 0; i < 2; i++ {
		for j := 0; j < 2; j++ {
			for k := 0; k < 2; k++ {
				weightV := NewVec3(u-float64(i), v-float64(j), w-float64(k))
				accum += (float64(i)*uu + float64(1-i)*(1-uu)) *
					(float64(j)*vv + float64(1-j)*(1-vv)) *
					(float64(k)*ww + float64(1-k)*(1-ww)) *
					Dot(&(*c)[i][j][k], weightV)
			}
		}
	}

	return accum
}

File: src/lib/perlin.rs
use crate::{dot, random_i32_in_range, unit_vector};
use crate::{Point3, Vec3};

const POINT_COUNT: usize = 256;

#[derive(Clone, Copy)]
pub struct Perlin {
    ranfloat: [Vec3; POINT_COUNT],
    perm_x: [i32; POINT_COUNT],
    perm_y: [i32; POINT_COUNT],
    perm_z: [i32; POINT_COUNT],
}

impl Perlin {
    pub fn new() -> Self {
        let mut ranfloat: [Vec3; POINT_COUNT] = [Vec3::default(); POINT_COUNT];

        for i in 0..POINT_COUNT {
            ranfloat[i] = unit_vector(&Vec3::random_in_range(-1., 1.));
        }

        Self {
            ranfloat,
            perm_x: Self::perlin_generate_perm(),
            perm_y: Self::perlin_generate_perm(),
            perm_z: Self::perlin_generate_perm(),
        }
    }

    #[inline]
    pub fn noise(&self, p: &Point3) -> f64 {
        let u = p.x() - p.x().floor();
        let v = p.y() - p.y().floor();
        let w = p.z() - p.z().floor();
        let i = p.x().floor() as i32;
        let j = p.y().floor() as i32;
        let k = p.z().floor() as i32;
        let mut c: [[[Vec3; 2]; 2]; 2] = [
            [[Vec3::default(); 2], [Vec3::default(); 2]],
            [[Vec3::default(); 2], [Vec3::default(); 2]],
        ];

        for di in 0..2 {
            for dj in 0..2 {
                for dk in 0..2 {
                    c[di][dj][dk] = self.ranfloat[(self.perm_x[((i + di as i32) & 255) as usize]
                        ^ self.perm_y[((j + dj as i32) & 255) as usize]
                        ^ self.perm_z[((k + dk as i32) & 255) as usize])
                        as usize];
                }
            }
        }

        Self::trilinear_interpolation(c, u, v, w)
    }

    #[inline]
    pub fn turbulence(&self, p: &Point3, depth: Option<i32>) -> f64 {
        let mut accum = 0.;
        let mut tmp_p = *p;
        let mut weight = 1.;

        for _ in 0..depth.unwrap_or(7) {
            accum += weight * self.noise(&tmp_p);
            weight *= 0.5;
            tmp_p *= 2.;
        }

        accum.abs()
    }

    #[inline]
    fn perlin_generate_perm() -> [i32; POINT_COUNT] {
        let mut p: [i32; POINT_COUNT] = [0; POINT_COUNT];

        for i in 0..POINT_COUNT {
            p[i] = i as i32;
        }

        Self::permute(p, POINT_COUNT);

        p
    }

    #[inline]
    fn permute(mut p: [i32; POINT_COUNT], n: usize) {
        for i in (1..n).rev() {
            let target = random_i32_in_range(0, i as i32);
            let tmp = p[i];

            p[i] = p[target as usize];
            p[target as usize] = tmp;
        }
    }

    #[inline]
    fn trilinear_interpolation(c: [[[Vec3; 2]; 2]; 2], u: f64, v: f64, w: f64) -> f64 {
        let uu = u.powi(2) * (3. - 2. * u);
        let vv = v.powi(2) * (3. - 2. * v);
        let ww = w.powi(2) * (3. - 2. * w);
        let mut accum = 0.0;

        for i in 0..2 {
            for j in 0..2 {
                for k in 0..2 {
                    let weight_v = Vec3::new(u - i as f64, v - j as f64, w - k as f64);
                    accum += (i as f64 * uu + (1 - i) as f64 * (1. - uu))
                        * (j as f64 * vv + (1 - j) as f64 * (1. - vv))
                        * (k as f64 * ww + (1 - k) as f64 * (1. - ww))
                        * dot(&c[i][j][k], &weight_v);
                }
            }
        }

        accum
    }
}
